opm/simulators/linalg/bda/WellContributionsOCLContainer.cpp

```diff
@@ -104,7 +116,7 @@ namespace bda
         }
 
         if(!wellContribs.multisegments.empty()){
-            multisegments = std::move(wellContribs.multisegments);
+            multisegments = std::move(std::make_unique<mswVecT>(wellContribs.multisegments));
```

Comment:
Same here.

Reply:
I still have the old version saved locally.

I began to study the rule of 5 but I'm having some difficulties in finding examples of it and, consequently, on how to apply it in our case here. If you guys could point me to some examples it would help a lot.

Sorry for not being able to contribute more in this matter.

Reply:
Disclaimer: I have not followed this closely myself - and my knowledge of the rule of 5 is mostly theoretical. But still: maybe if you describe your problem with words it would be simpler for us to assist you?

Reply:
I hadn't heard of the rule of 5 until yesterday and apparently my old implementation (before the last 2 commits) was disrespecting this rule and, therefore, was considered bad practice. So my problem is that I still don't know how and where to put the copy/move constructors and destructors. Should I put them in the `MultisegmentWell` class or in the `WellContributionsOCLContainer` class? Should these destructors and constructors operate in the elements of the vector (which are complex objects by themselves) or should they operate on the vector as whole? These are the questions I'm struggling with. But as I said, I've heard of all this yesterday and only today I'm grasping what it all means. I think I'll create another PR with my attempt at implementing it in this case but I don't know how soon I'll be able to do it

Reply:
If you have the old version, the simplest may be to go back that, and explicitly delete the copy constructor and copy assignment operator in line with @blattms's suggestion, instead of introducing smart pointers. That would solve the problem of these operations being wrong.

I looked at the modifications done in the "smart pointer" commit, and became somewhat confused. What I intended was that instead of having a vector of pointers, you should have a vector of smart pointer. The change in that commit was something different, it now uses a smart pointer to a vector of pointers, which saves no cleanup. The "problem" is the plain pointers owning the resources they point to. I have now read a bit more of the code, and see that these pointers are not created in this class, but moved from another object, `wellContribs.multisegments`,  (that presumably has the same kind of issue). I think the best solution could be to refactor that class to using a `vector<unique_ptr<MultisegmentWellContribution>>` instead of a vector of plain pointers. Then the rest of the code here would remain unchanged except for that type. I do not know if this is feasible, as I have not looked at that class' code.

If pressed for time, do the simple solution (the first paragraph). In the long run, the second may save you some pain.

Reply:
@ducbueno Please don't try the refactoring. If somebody insist on it that can be done later. You have enough on your plate and I am very grateful for your persistent bug hunting (which most of the time are not even due to you).

`wellContribs.multisegments` is used in CUDA code and that does not have the full C++ functionality. Maybe unique-ptr is even not fully supported there. Hence sticking with simple C for its entries might make sense,

Reply:
> So my problem is that I still don't know how and where to put the copy/move constructors and destructors. Should I put them in the `MultisegmentWell` class or in the `WellContributionsOCLContainer` class? Should these destructors and constructors operate in the elements of the vector (which are complex objects by themselves) or should they operate on the vector as whole?

The goal is to avoid writing any of the 5 functions (including destructors). You can accomplish that by using appropriate smart pointers, because then the compiler-generated default versions of the 5 functions will be correct.

The problem is that when a plain pointer is destroyed, what it points to is not `delete`d. So if you have a vector of plain pointers, when the vector dies it will deallocate the array where the pointers were stored without calling `delete` on them. Therefore, you did the right thing to explicitly call `delete` in the destructors. However, this is error prone, since you have to make the other of the 5 functions for this to be correct (or, explicitly delete those functions).

By replacing the plain pointers of the `multisegment` member in the WellContributions class with std::unique_ptr, you accomplish the following: you do not have to `delete` the pointers anymore, simplifying your WellContributions destructor. (If you had implemented all 5 of the set, you would see similar simplifications in all of them.) Even better, the other class that "steals" these pointers by moving them into its own vector, would also no longer need to delete them. (This was the place that triggered the original comments).

Two more things:
 - In the WellContributions class there are several more members handled by explicit pointers, which prevents you from removing the destructor in that class. I think that all of these can be replaced with std::vector instead. (Using std::vector is better than using a smart pointer to an array, since it has many useful methods, and everyone knows how to use it.) I am not sure what the freeCudaMemory() call does, you may still need to do that in the destructor for all I know.
 - If you use std::unique_ptr in a class, you will not be able to copy objects of that class (assuming you do just let the compiler implement the default versions for the 5 functions). Two of the 5, the copy constructor and copy assignment operator, will not be available. This is because it is not allowed to copy std::unique_ptr. Often, this is exactly what you want! It is possible to move a unique_ptr, so that is still allowed for the class containing it.

Reply:
@blattms please check the last commit and see if everything is correct. The `BASE1_MSW_HFA.DATA` is now running again.

@atgeirr thank you for the very complete answer, it made things a lot clearer. I'll experiment with the smart pointers in the way you suggested in a new branch for a while. However I may not have the necessary time to refactor things fully, since as @blattms the `wellContribs` object is also used by CUDA and I'll have to be extra careful for not breaking that and I also have to give some attention to https://github.com/OPM/opm-simulators/issues/2755, as it is related to my PhD project.

Sorry for causing all this discussion and thank you again. At least I was able to learn something :)

Reply:
> At least I was able to learn something :)

Then it was worthwhile, I guess!